### PR TITLE
Add classifier → ml_signal backtest example

### PR DESCRIPTION
**Pull request by jrmeier**
## Summary
Adds a practical path to try a return classifier with fast-trade’s existing backtest engine, with review fixes for OOS correctness:

- `fast_trade/ml/classifier.py` — features, forward-return labels, `HistGradientBoostingClassifier` fit, `ml_signal` attach, and `run_classifier_backtest()`
- **Purge gap**: last `horizon` train rows dropped so forward labels do not leak into the holdout
- **Freq guard**: infer strategy freq from the frame; reject silent resample mismatches (`1h`/`h` aliases OK); `allow_resample=True` to opt in
- **Probability threshold**: `predict_proba` + configurable `signal_threshold` (CLI `--signal-threshold`)
- `column` datapoint transformer for precomputed signals (docs note resample `.first()` behavior)
- Explicit `scikit-learn` dependency
- Example script + strategy YAML shape + unit tests (including edge paths for 100% coverage)

## How to try
```bash
python examples/ml_classifier_backtest.py --synthetic
python examples/ml_classifier_backtest.py --synthetic --signal-threshold 0.55
# or with archive data:
# ft download BTCUSDT binanceus --start 2024-01-01 --end 2025-01-01
# python examples/ml_classifier_backtest.py --symbol BTCUSDT --exchange binanceus
```

## Testing
- `python -m pytest test/test_classifier.py` (18 passed; classifier.py at 100% coverage)
- `flake8` on touched Python files
- Manual synthetic run (shows purged train rows + inferred freq)

## CI fix
Build matrix failed on `fail-under=100` because 4 statements in `classifier.py` were uncovered (purge validation, uninferable freq, predict fallback). Those paths are now tested.

<div><a href="https://cursor.com/agents/bc-01a02719-fcf3-74f2-bb14-27a971e9639f?cursor_ref=pr_footer&cursor_cta=open_in_web"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/open-in-web-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/open-in-web-light.png">Open in Web</picture></a>&nbsp;<a href="https://cursor.com/background-agent?bcId=bc-01a02719-fcf3-74f2-bb14-27a971e9639f&cursor_ref=pr_footer&cursor_cta=open_in_cursor"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/open-in-cursor-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/open-in-cursor-light.png">Open in Cursor</picture></a>&nbsp;</div>

### CODE REVIEW

File: fast_trade/ml/classifier.py [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+    x_train = usable.loc[train_idx, cols]
+    y_train = usable.loc[train_idx, "y"].astype(int)
+    x_test = usable.loc[test_idx, cols]
+    y_test = usable.loc[test_idx, "y"].astype(int)
+
+    if y_train.nunique() < 2:
+        raise ValueError("Training labels must include both classes; loosen threshold")
+
+    model = HistGradientBoostingClassifier(random_state=random_state)
+    model.fit(x_train, y_train)
```
Comment (jrmeier) [inline fast_trade/ml/classifier.py]: **High — training labels leak into the holdout window**

`label_forward_return` uses `close.shift(-horizon)`, then `_time_split_index` cuts on the usable (already-labeled) index. The last `horizon` training rows have targets that depend on closes inside the test window, so `backtest_on="test"` is not a clean OOS split and holdout accuracy/AUC can be optimistically biased.

**Fix:** after computing the cut, drop the last `horizon` rows from `train_idx` (embargo/purge gap), or build labels only from within-train future prices. Also worth asserting in a unit test that `max(train_idx) + horizon < min(test_idx)` in time.

File: fast_trade/ml/classifier.py [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+    comission: float = 0.01,
+    **overrides: Any,
+) -> Dict[str, Any]:
+    """YAML-shaped strategy that enters when the classifier predicts 1."""
+    strategy: Dict[str, Any] = {
+        "name": "ml_classifier_example",
+        "symbol": symbol,
+        "exchange": exchange,
+        "freq": freq,
+        "comission": comission,
```
Comment (jrmeier) [inline fast_trade/ml/classifier.py:331]: **Medium — default `freq` can silently desync prices and `ml_signal`**

`default_classifier_strategy` hardcodes `freq="1H"`. Callers that pass a dataframe at another bar size without `strategy_overrides["freq"]` still go through `prepare_df` → `resample(freq).first()`, which regrids both OHLCV and the precomputed signal. Results look plausible but are wrong.

The example script does override freq (good); the library helper does not infer or validate that strategy freq matches `df`’s native spacing. Prefer inferring from `df.index`, requiring an explicit freq, or failing when they disagree.

File: fast_trade/ml/classifier.py [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+def predict_ml_signal(
+    model: Any,
+    features: pd.DataFrame,
+    feature_columns: Sequence[str],
+) -> pd.Series:
+    """Return a 0/1 signal series aligned to ``features`` index."""
+    cols = list(feature_columns)
+    x = features[cols]
+    pred = model.predict(x)
+    return pd.Series(pred.astype(int), index=features.index, name="ml_signal")
```
Comment (jrmeier) [inline fast_trade/ml/classifier.py:305]: **Medium — hard `predict()` for a trading signal**

With a positive return threshold, class imbalance usually favors `0`. `predict()` uses the model’s default decision threshold, so the strategy can spend most bars flat (or rarely enter) even when probabilities are informative. For a demo meant to show ML → backtest wiring, exposing `predict_proba` + a configurable probability cutoff (default `0.5`) would make the enter/exit behavior much more controllable and easier to reason about.

File: fast_trade/transformers_map.py [hunk 4]
```diff
@@ -1,10 +1,29 @@
...
+
+    Useful for ML signals (or any external series) that are attached before
+    ``run_backtest`` / ``prepare_df``.
+    """
+    if column not in df.columns:
+        raise ValueError(
+            f"Column '{column}' not found in dataframe. "
+            "Attach it before running the backtest."
+        )
+    return df[column]
```
Comment (jrmeier) [inline fast_trade/transformers_map.py:26]: **Low — `column` is a useful escape hatch, but document the resample interaction**

`prepare_df` resamples *before* transformers run (`resample(...).first()` on every column). Precomputed `ml_signal` therefore survives only as “first value in each bucket.” That is fine when strategy `freq` matches the frame; when downsampling it silently changes signal timing. A one-line note in this docstring (and in the strategy YAML) would save future users.

File: fast_trade/ml/classifier.py [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from typing import Any, Dict, List, Mapping, Optional, Sequence, Tuple
+
+import numpy as np
+import pandas as pd
+from sklearn.ensemble import HistGradientBoostingClassifier
+from sklearn.metrics import accuracy_score, roc_auc_score
```
Comment (jrmeier) [inline fast_trade/ml/classifier.py:22]: **Low — packaging**

This module hard-imports `sklearn`, but `pyproject.toml` does not declare `scikit-learn` (only `hmmlearn`, which happens to depend on it). Same latent issue already exists in `hmm_screen.py`. Worth adding an explicit `scikit-learn` dependency (or a clear optional extra) so a minimal install does not break on import.

**Review by jrmeier — COMMENTED** [review]
## Deep review — classifier → `ml_signal` example (#45)

Solid, readable demo path: features → labels → fit → attach column → existing enter/exit engine. The `column` transformer is the right integration seam, tests cover the happy path and most validation errors, and docs point people at the runnable example.

### Blockers / should-fix before treating metrics as OOS

1. **Training-label leakage across the split (high)** — forward labels use `shift(-horizon)`, then the time split is applied on already-labeled rows. The last `horizon` train targets peek into the holdout. Purge/embargo those rows (or rebuild labels inside the train window only).
2. **Default strategy `freq="1H"` (medium)** — library callers who omit `strategy_overrides` can get silent resample desync between prices and `ml_signal`. Infer, require, or validate freq against the frame.

### Design / product notes (non-blocking)

3. Hard `predict()` + enter-on-1 / exit-on-0 is brittle under class imbalance; a probability threshold would make the demo more honest.
4. Same-bar execution uses features built from the bar’s close and enters on that bar — consistent only if fills are at close after the close is known; worth one sentence in the module/example docstring.
5. `backtest_on="all"` is correctly labeled leaky; good.
6. Declare `scikit-learn` explicitly in packaging (currently transitive via `hmmlearn`).
7. Strategy YAML is documentation-only — fine, but `ft backtest` on it alone will fail without a pre-attached column; the comment already says so.

### What’s in good shape

- Clear API surface (`fit_return_classifier`, `attach_ml_signal`, `run_classifier_backtest`)
- Holdout-first default (`backtest_on="test"`)
- Example script correctly passes `freq` through overrides
- Unit tests for features/labels, column transformer, validation, and end-to-end synthetic backtest

Happy to follow up with a purge-gap + freq-guard patch if you want that landed on this branch.

**Comment by jrmeier** [comment]
Addressed the review findings on this branch (`10d038e`):

1. **Label leakage** — `_time_split_index(..., purge_bars=horizon)` drops the last `horizon` train rows; unit test asserts `max(train_pos)+horizon < test_start`.
2. **Freq desync** — `resolve_backtest_freq` infers from the frame, treats `1h`/`h` as equal via `to_offset`, and raises on mismatch unless `allow_resample=True`.
3. **Hard predict** — `predict_ml_signal` uses `predict_proba` with configurable `signal_threshold` (example: `--signal-threshold`).
4. **Docs** — same-bar close-fill note on the module; `column` transformer + strategy YAML warn about `resample(...).first()`.
5. **Packaging** — `scikit-learn>=1.5.0` added to `pyproject.toml`.

`pytest test/test_classifier.py` → 15 passed.
